src/import/load-costume.js

```diff
@@ -54,123 +101,114 @@ const fetchBitmapCanvas_ = function (costume, runtime, rotationCenter) {
         return Promise.reject('No V2 Bitmap adapter present.');
     }
 
-    return new Promise((resolve, reject) => {
-        const baseImageElement = new Image();
-        let textImageElement;
+    return Promise.all([costume.asset, costume.textLayerAsset].map(asset => {
+        if (!asset) {
+            return null;
+        }
 
-        // We need to wait for 2 images total to load. loadedOne will be true when one
-        // is done, and we are just waiting for one more.
-        let loadedOne = false;
+        if (typeof createImageBitmap !== 'undefined') {
+            return createImageBitmap(
+                new Blob([asset.data], {type: asset.assetType.contentType})
+            );
+        }
 
-        const onError = function () {
-            // eslint-disable-next-line no-use-before-define
-            removeEventListeners();
-            reject('Costume load failed. Asset could not be read.');
-        };
-        const onLoad = function () {
-            if (loadedOne) {
-                // eslint-disable-next-line no-use-before-define
-                removeEventListeners();
-                resolve([baseImageElement, textImageElement]);
-            } else {
-                loadedOne = true;
-            }
-        };
+        return new Promise((resolve, reject) => {
+            const image = new Image();
+            image.onload = function () {
+                resolve(image);
+                image.onload = null;
+                image.onerror = null;
+            };
+            image.onerror = function () {
+                reject('Costume load failed. Asset could not be read.');
+                image.onload = null;
+                image.onerror = null;
+            };
+            image.src = asset.encodeDataURI();
+        });
+    }))
+        .then(([baseImageElement, textImageElement]) => {
+            const mergeCanvas = canvasPool.create();
+
+            const scale = costume.bitmapResolution === 1 ? 2 : 1;
+            mergeCanvas.width = baseImageElement.width;
+            mergeCanvas.height = baseImageElement.height;
 
-        const removeEventListeners = function () {
-            baseImageElement.removeEventListener('error', onError);
-            baseImageElement.removeEventListener('load', onLoad);
+            const ctx = mergeCanvas.getContext('2d');
+            ctx.drawImage(baseImageElement, 0, 0);
             if (textImageElement) {
-                textImageElement.removeEventListener('error', onError);
-                textImageElement.removeEventListener('load', onLoad);
+                ctx.drawImage(textImageElement, 0, 0);
+            }
+            let canvas = mergeCanvas;
```

Comment:
Why do we need this assignment (also why are both `canvas` and `mergeCanvas` being returned below)?

Reply:
Because of the line, two lines down.

```js
canvas = runtime.v2BitmapAdapter.resize(mergeCanvas, canvas.width * scale, canvas.height * scale);
```

v2BitmapAdapter returns a different canvas than the one passed in. So we cannot know the `mergeCanvas` value once `canvas` is overwritten if we only passed `canvas` to the next section.

```js
canvasPool.release(mergeCanvas);
```

In other words `canvas` is not guaranteed to be `mergeCanvas`. `mergeCanvas` is the one we want to return to the resource pool.

Reply:
Why does it matter which canvas we return to the resource pool?

If we never had this line, and then instead in the conditional two lines down, we had:

```mergeCanvas = runtime.v2BitmapAdapter.resize(mergeCanvas, mergeCanvas.width * scale, mergeCanvas.height * scale);```

Wouldn't the old canvas get garbage collected?

Reply:
I figured I'd use ownership semantics here. I am making the assumptions mergeCanvas is owned by this file and functions and that the canvas returned by resize is owned by resize. Since resize borrows mergeCanvas and this file borrows resize's canvas, we should only hold on to mergeCanvas. Since we don't own resize's canvas, we shouldn't retain it and later transform it by merging bitmaps on it.

The point is to keep canvases from being garbage collected. Under this formality we can only keep the mergeCanvas.

Reply:
That makes sense to me, thanks for explaining. Do you mind adding a comment to the code so that this is called out explicitly as I imagine this question would come up again and the code could get rewritten in a way that isn't cohesive with this thought process.

Reply:
Sure thing.